compiler/optimizing/induction_var_simplification.cc

```diff
@@ -0,0 +1,318 @@
+       // check if it is incremented by a constant amount 
+       ArenaSet<HInstruction*>* set = induction_analysis_->LookupCycle(phi);
+       for (HInstruction* ins : *set) {
+          if (ins->IsAdd()) { 
+            HBinaryOperation* op = ins->AsAdd();
+            HConstant* cst = op->GetConstantRight();
+            if (cst != nullptr) {
+               biv_increment_=op;
+              return true;
+            }
+        }
+      }
+      }
+    }
+  }
+  return false;
+}
+
+void InductionVarSimplification::PerformReduction(HLoopInformation* loop, HInstruction* derived_var,
+                  std::tuple<HInstruction*, HConstant*, HConstant*> val) {
+    DCHECK(biv_increment_ != nullptr);
+    HInstruction* var = std::get<0>(val);
+    HPhi* phi = var->AsPhi();
+    DCHECK(phi != nullptr);
+    HInstruction* input = GetOutOfLoopInput(loop, phi);
+    // Create two new instructions mul and add
+    ArenaAllocator* allocator = derived_var->GetBlock()->GetGraph()->GetAllocator();
+    HInstruction* new_mul =  new (allocator) HMul(derived_var->GetType(), input, std::get<1>(val), derived_var->GetDexPc());
+    loop->GetPreHeader()->InsertInstructionBefore(new_mul, loop->GetPreHeader()->GetLastInstruction());
+    HInstruction* new_add = new (allocator) HAdd(new_mul->GetType(), new_mul,std::get<2>(val), derived_var->GetDexPc());
+    loop->GetPreHeader()->InsertInstructionBefore(new_add, loop->GetPreHeader()->GetLastInstruction());
+    HPhi* new_phi = new (allocator) HPhi(allocator, kNoRegNumber, 0, phi->GetType());
+    phi->GetBlock()->InsertPhiAfter(new_phi, phi);
+    new_phi->AddInput(new_add);
+    HConstant* biv_inc_cst = biv_increment_->GetConstantRight();
+    HConstant* biv_cst = GetNewConstant(HInstruction::InstructionKind::kMul, biv_inc_cst, std::get<1>(val));
+    DCHECK(biv_cst != nullptr);
+    HAdd* biv_add = new (allocator) HAdd(biv_inc_cst->GetType(), new_phi, biv_cst);
+    biv_increment_->GetBlock()->InsertInstructionAfter(biv_add, biv_increment_);
+    new_phi->AddInput(biv_add);
+    // Replace uses of new add within the loop with new_phi
+    const HUseList<HInstruction*>& uses = derived_var->GetUses();
+    for (auto it = uses.begin(), end = uses.end(); it != end;) {
+      HInstruction* user = it->GetUser();
+      size_t input_index = it->GetIndex();
+      ++it;
+      user->ReplaceInput(new_phi, input_index);
+    }
+    const HUseList<HEnvironment*>& env_uses = derived_var->GetEnvUses();
+    for (auto it = env_uses.begin(), end = env_uses.end(); it != end;) {
+      HEnvironment* user = it->GetUser();
+      size_t index = it->GetIndex();
+      ++it;  // increment prior to potential removal
+      user->RemoveAsUserOfInput(index);
+      user->SetRawEnvAt(index, new_phi);
+      new_phi->AddEnvUseAt(user, index);
+    } 
+    DCHECK(!derived_var->HasUses());
+}
+
+bool InductionVarSimplification::IsCandidateForReduction (
```

Comment:
Oh I see that this one is actually identifying the reduction candidates. Then what's exactly the purpose of IsCandidatePhi ?